### PR TITLE
7.0 web advanced search x2x

### PR DESCRIPTION
**Pull request by gfcapalbo**
Backporting web_advanced_search to 7.0, very little to do, just putting the js in the manifest.
Small JS change, because of the different layout of the search area in 7, this feature was unusable.
Every time you would select a criteria the search would close, forcing the user to reopen it and confirm.
Now the search interface remains always open until the user clicks to close it.

### CODE REVIEW

**Comment by gfcapalbo** [comment]
Hello, just added a js modification that prohibits the searchview from closing. This makes the module very usable in 7.0 interface.

File: web_advanced_search_x2x/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#
+##############################################################################
+{
+    "name": "Search x2x fields",
+    "version": "1.0",
+    "author": "Therp BV, "
+              "Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "Usability",
+    "summary": "Use a search widget in advanced search for x2x fields",
```
Comment (pedrobaeza) [inline web_advanced_search_x2x/__openerp__.py]: You have to put the description also here, because in v7 README.rst is not parsed

**Comment by pedrobaeza** [comment]
One minor remark. The rest works very well.

**Comment by pedrobaeza** [comment]
:+1: 

Please check CLA question

**Comment by StefanRijnhart** [comment]
Now added @gfcapalbo as a contact of Therp in the OCA instance. Therp's ECLA should cover him now.

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+                    });
+                    result.domains[index] = combined;
+                }
+            });
+            return result;
+        },
+    })
+
+
+    instance.web.PermanentSearchView = instance.web.SearchView.include({
```
Comment (hbrunn) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: you want to modify the existing SearchView, so there's no need to keep the reference to the updated object (which is the same as `instance.web.SearchView`)

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+                       curr_class = ev.target.attributes.class.value;
+                       //if the class is one of the search popups defined in this module, please don't close the search view when using them.
+                       if ((curr_class == "oe_button oe_selectcreatepopup-search-select oe_highlight") || 
+                           (curr_class == "oe_highlight oe_selectcreatepopup-search-select-domain")    ||
+                           (curr_class == "oe_button oe_selectcreatepopup-search-close oe_bold oe_form_button_cancel") ||
+                           (curr_class == "ui-dialog-titlebar-close ui-corner-all"))
+                           {
+                            if (this.$el)
+                                {
+                                 this.$el.addClass('oe_searchview_open_drawer');
```
Comment (hbrunn) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: wouldn't is be a lot simpler to add those classes in `instance.web_advanced_search_x2x.SelectCreatePopup`?

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+                           }
+                      }
+                  //again, check if there are two parentelements, attributes and  a class , avoid crashes
+                  if (ev.target &&  ev.target.parentElement &&  ev.target.parentElement.parentElement &&  ev.target.parentElement.parentElement.attributes)
+                      {
+                       if ((ev.target.parentElement.parentElement.attributes.class) && (this.el.className))
+                           { //if the target element of the event is a search or a create or anything from the the autocomplete AND the element is part of a search view, also please don't close the searchview.
+                            if ((ev.target.parentElement.parentElement.attributes.class.value == "ui-autocomplete ui-menu ui-widget ui-widget-content ui-corner-all openerp") && (this.el.className.indexOf("oe_searchview")!= -1 ))
+                               {
+                                this.$el.addClass('oe_searchview_open_drawer');
```
Comment (hbrunn) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: this should be doable in `create_searchfield` above

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+                               }
+                            }
+                      }
+                  })
+                  return $.when(p,this.ready);
```
Comment (hbrunn) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: `_super` already checks for `this.ready`, so returning `p` suffices

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        start: function()
+              {   var self = this;
+                  var p = this._super();
+                  instance.web.bus.on('click', this, function(ev) {
+                  //check if there is a class, avoid crashes
+                  if (( ev.target.attributes.class) &&  (ev.target.attributes))
+                      {
+                       curr_class = ev.target.attributes.class.value;
+                       //if the class is one of the search popups defined in this module, please don't close the search view when using them.
+                       if ((curr_class == "oe_button oe_selectcreatepopup-search-select oe_highlight") ||
```
Comment (hbrunn) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: you don't really need that as pointed out below, but for the future: Use `jQuery(ev.target).has_class()` to check for classes. Or `jQuery(ev.target).filter('.all.the.classes.to.check.for').length`. Both of them have the advantage that you don't need the checks above and more importantly, they are insensitive to changes in order and additions

File: web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js (hunk 4 above)
Comment (gfcapalbo) [inline web_advanced_search_x2x/static/src/js/web_advanced_search_x2x.js]: this second case was created to cover cases in which the user would resize or click on other elements, activating "click" event but without a target.attribute, I don't see how can i include that in the create_searchfield. 
(Merged a redundant if)

**Comment by hbrunn** [comment]
@gfcapalbo did you look into https://github.com/OCA/web/pull/148#discussion_r33225251 ?

**Comment by legalsylvain** [comment]
@gfcapalbo : any news ?

**Comment by hbrunn** [comment]
@gfcapalbo don't forget to cherrypick (and probably backport) the evolution that happened in 8.0 meanwhile

**Comment by hbrunn** [comment]
@gfcapalbo ping

**Comment by hbrunn** [comment]
don't forget to cherry-pick the fixes done on the 8.0 branch

**Comment by andreschenkels** [comment]
Works here :thumbsup:

**Comment by gfcapalbo** [comment]
tested and ported to 7. Incorporated improvements from 8.0. commits from 8 all squashed in  
5491730. references in commit message.

File: web_advanced_search_x2x/README.rst [hunk 7]
```diff
@@ -0,0 +1,53 @@
+Search for x2x records in advanced search
```
Comment (hbrunn) [inline web_advanced_search_x2x/README.rst]: remove this file, it doesn't do any good in v7

File: web_advanced_search_x2x/__init__.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Comment (hbrunn) [inline web_advanced_search_x2x/__init__.py]: take the opportunity to use short license headers everywhere

**Comment by hbrunn** [comment]
a few nitpicks, and can you do something about

screenshot_20160914_190357

**Comment by hbrunn** [comment]
(this is from searching modules on runbot)

**Comment by gfcapalbo** [comment]
I noticed that failed to catch it will try again.

**Comment by andreschenkels** [comment]
Can this be reviewed? Or is it still being developped?

**Comment by hbrunn** [comment]
@gfcapalbo needs to find time to fix the bugs mentioned already, but if you find new ones, this will be appreciated for sure

**Review by hbrunn — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Please squash commit history by logical set/author and I'll merge this one.

**Comment by gfcapalbo** [comment]
@pedrobaeza  i have rebased to all my commits to the maximum level  (there are only two commits of mine they are separate and non contiguous ) , anything further would conceal commits from @hbrunn and @legalsylvain  (commits from cherry-picking 8.0).

**Comment by pedrobaeza** [comment]
@hbrunn do you mind squashing your commits too? or may I? @gfcapalbo now Travis is red. Do you know why?

**Comment by hbrunn** [comment]
travis is red because of https://travis-ci.org/OCA/web/jobs/373015996#L579

you may smash whatever you want on merging, but then I don't see the point of first forcing contributors to the process of getting commits from the other branch just to squash them together again

**Comment by pedrobaeza** [comment]
No, I'm not talking about squash on merging, but checkout your branch, squash your commits, and push again to this PR. This is for not making you lose time for a nitpicking.

I see the error, that is unrelated and easily fixable. Doing on main branch via a direct commit.

**Comment by pedrobaeza** [comment]
Merged in main branch as I wasn't able to push to this one.

**Pull request closed without merging**